src/s3/hash.c

```diff
@@ -0,0 +1,245 @@
+								filename, buf.data)));
+
+			/* filename is already filled in */
+			strlcpy(newEntry->hash, fileEntry.hash, sizeof(fileEntry.hash));
+			newEntry->checkpointNumber = fileEntry.checkpointNumber;
+			newEntry->changed = false;
+		}
+		else
+			ereport(ERROR,
+					(errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+					 errmsg("invalid line format of the hash file \"%s\": %s",
+							filename, buf.data)));
+	}
+
+	pfree(buf.data);
+
+	if (ferror(file))
+		ereport(ERROR,
+				(errcode_for_file_access(),
+				 errmsg("could not read file \"%s\": %m", filename)));
+
+	FreeFile(file);
+}
+
+/*
+ * Save current workers' PGDATA files hash into a temporary file.
+ */
+void
+flushS3PGFilesHash(S3HashState *state, const char *filename)
+{
+	FILE	   *file;
+
+	Assert(state->pgFiles != NULL);
+
+	/*
+	 * We open the file for append in case if previous flush already created
+	 * the file.
+	 */
+	file = AllocateFile(filename, "a");
+	if (file == NULL)
+		ereport(ERROR,
+				(errcode_for_file_access(),
+				 errmsg("could not create file \"%s\": %m", filename)));
+
+	for (int i = 0; i < state->pgFilesLen; i++)
+	{
+		if (fprintf(file, "FILE: %s, HASH: %s, CHECKPOINT: %u\n",
+					state->pgFiles[i].filename, state->pgFiles[i].hash,
+					state->pgFiles[i].checkpointNumber) < 0)
+			ereport(ERROR,
+					(errcode_for_file_access(),
+					 errmsg("could not write file \"%s\": %m",
+							filename)));
+	}
+
+	if (fflush(file) || ferror(file) || FreeFile(file))
+		ereport(ERROR,
+				(errcode_for_file_access(),
+				 errmsg("could not write file \"%s\": %m",
+						filename)));
+
+	state->pgFilesLen = 0;
+}
+
+/*
+ * Check if a PGDATA file changed since last checkpoint and return PGFileHash.
+ */
+S3FileHash *
+getS3PGFileHash(S3HashState *state, const char *filename,
+				Pointer data, uint64 size)
+{
+	S3FileHash *prevEntry = NULL;
+	S3FileHash *newEntry;
+	unsigned char hashbuf[PG_SHA256_DIGEST_LENGTH];
+	char		hashstringbuf[PG_SHA256_DIGEST_STRING_LENGTH];
+
+	/*
+	 * hashTable might not have been initialized before if a hash file hasn't
+	 * been found.
+	 */
+	if (state->hashTable != NULL)
+	{
+		char		key[MAXPGPATH];
+
+		MemSet(key, 0, sizeof(key));
+		strlcpy(key, filename, MAXPGPATH);
+
+		prevEntry = (S3FileHash *) hash_search(state->hashTable, key,
+											   HASH_FIND, NULL);
+	}
+
+	(void) SHA256((unsigned char *) data, size, hashbuf);
+
+	hex_encode((char *) hashbuf, sizeof(hashbuf), hashstringbuf);
+	hashstringbuf[PG_SHA256_DIGEST_STRING_LENGTH - 1] = '\0';
+
+	newEntry = (S3FileHash *) palloc0(sizeof(S3FileHash));
+
+	strlcpy(newEntry->filename, filename, sizeof(newEntry->filename));
+	strlcpy(newEntry->hash, hashstringbuf, sizeof(newEntry->hash));
+
+	newEntry->changed = ((prevEntry == NULL) ||
+						 (strncmp(prevEntry->hash, newEntry->hash, sizeof(newEntry->hash)) != 0)) ?
+		1 : 0;
+	newEntry->checkpointNumber = (newEntry->changed) ?
+		state->checkpointNumber : prevEntry->checkpointNumber;
```

Comment:
Maybe split this block to if(prevEntry == NULL){} else {} to increase readability and to be clear that we never dereference NULL ptr (in fact we don't) ?

Reply:
I rewrote to make it more readable.